File: src/dals/UserDal.ts
import { Pool } from 'pg';
import { UserModel } from '../models/UserModel';
import { DataBase } from '../config/DataBase'; // Singleton de conexão

export class UserDAL 
{
  private db: Pool;

  constructor()
  {
    this.db = DataBase.getInstance(); // Usa Singleton para o Pool
  }

  //buscando por um usuario pelo cpf
  async findByCPF(cpf: string): Promise<UserModel | null>
  {
    const result = await this.db.query('SELECT * FROM credenciais WHERE cre_login = $1', [cpf]);
    if(result.rows.length != 0)
    {
        const row = result.rows[0];
        return new UserModel(row.cre_login,row.cre_senha,row.cre_cod);
    } 
    return null;
  }

  async create(user: UserModel): Promise<UserModel>
  {
    const result = await this.db.query(
        'INSERT INTO credenciais (cre_login, cre_senha) VALUES ($1, $2) RETURNING cre_cod',
        [user.getCpf(), user.getSenha()]
    );
    const insertedId = result.rows[0].cre_cod;
    return new UserModel(user.getCpf(), user.getSenha(), insertedId);
  }

  //update
  async updateSenha(cpf: string, novaSenha: string): Promise<void>
  {
    await this.db.query('UPDATE credenciais SET cre_senha = $1 WHERE cre_login = $2',[novaSenha, cpf]);
  }

  //delete
  async deleteByCPF(cpf: string): Promise<void>
  {
    await this.db.query('DELETE FROM credenciais WHERE cre_login = $1', [cpf]);
  }
}

File: BackEnd/src/main/java/SoftPet/backend/dal/UserDAL.java
package SoftPet.backend.dal;

import SoftPet.backend.config.SingletonDB;
import SoftPet.backend.model.UserModel;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class UserDAL {

    public UserModel findByCPF(String cpf)
    {
        UserModel user = null;
        String sql = "SELECT * FROM credenciais WHERE cre_login = ?";
        try(PreparedStatement stmt = SingletonDB.getConexao().getPreparedStatement(sql))
        {
            stmt.setString(1, cpf);
            ResultSet rs = stmt.executeQuery();
            if (rs.next())
            {
                user = new UserModel(
                        rs.getLong("cre_cod"),
                        rs.getString("cre_login"),
                        rs.getString("cre_senha")
                );
            }
        }catch(SQLException e)
        {
            e.printStackTrace();
        }
        return user;
    }

    public UserModel create(UserModel user) {
        String sql = "INSERT INTO credenciais (cre_login, cre_senha) VALUES (?, ?)";
        try (PreparedStatement stmt = SingletonDB.getConexao().getPreparedStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            stmt.setString(1, user.getCpf());
            stmt.setString(2, user.getSenha());
            int affectedRows = stmt.executeUpdate();

            if (affectedRows > 0) {
                ResultSet rs = stmt.getGeneratedKeys();
                if (rs.next()) {
                    user.setId(rs.getLong(1));
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return user;
    }

    public boolean updateSenha(String cpf, String novaSenha) {
        String sql = "UPDATE credenciais SET cre_senha = ? WHERE cre_login = ?";
        try (PreparedStatement stmt = SingletonDB.getConexao().getPreparedStatement(sql)) {
            stmt.setString(1, novaSenha);
            stmt.setString(2, cpf);
            return stmt.executeUpdate() > 0;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    public boolean deleteByCPF(String cpf) {
        String sql = "DELETE FROM credenciais WHERE cre_login = ?";
        try (PreparedStatement stmt = SingletonDB.getConexao().getPreparedStatement(sql)) {
            stmt.setString(1, cpf);
            return stmt.executeUpdate() > 0;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    // Métodos antigos opcionais (caso você ainda use a interface IDAL)
    public List<UserModel> getAll() {
        List<UserModel> list = new ArrayList<>();
        String sql = "SELECT * FROM credenciais";
        ResultSet rs = SingletonDB.getConexao().consultar(sql);
        try {
            while (rs.next()) {
                UserModel user = new UserModel(
                        rs.getLong("cre_cod"),
                        rs.getString("cre_login"),
                        rs.getString("cre_senha")
                );
                list.add(user);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return list;
    }
}
